text/0000-build-std/6-stage-2.md

```diff
@@ -0,0 +1,176 @@
+# Stage 2: `build-std=compatible`
+
+This stage proposes extending the `build-std` option with a new `compatible`
+value, which will become the default and automatically rebuild the standard
+library when it is necessary to maintain compatibility with the compiler flags
+used by the rest of the crate graph.
+
+This is aimed at unblocking the stabilisation of ABI-modifying compiler flags
+(as per [motivations](./3-motivation.md)).
+
+# Proposal
+[proposal]: #proposal
+
+The `build-std` option in the Cargo configuration will be extended with a new
+value:
+
+```toml
+[build]
+build-std = "compatible" # or `always`/`never`
+```
+
+"compatible" will become the default value for `build-std` ([?][rationale-default]).
+
+When `build-std` is set to "compatible", then the standard library crates will
+be rebuilt automatically ([?][rationale-why-automatic]) when a pre-built
+standard library is not present or the pre-built standard library is
+incompatible with the rest of the crate (due to use of target modifiers).
+
+> [!NOTE]
+>
+> rustc will be extended with the ability to determine whether an existing
+> `rlib` artefact is compatible with the flags passed to rustc
+> ([?][rationale-rustc-support]).
+>
+> ```shell-session
+> $ rustc -Zreg-struct-return core/lib.rs -o libcore.rlib
+> $ rustc --emit compatibility $cargo_flags libcore.rlib
+> error: mixing `-Zreg-struct-return` will cause an ABI mismatch in crate `core`
+> ```
+>
+> In the above example, rustc compiles libcore with the `-Zreg-struct-return`
+> target modifier to create an `rlib`, and then is invoked to check the
+> compatibility of the flags Cargo would have passed to the crate against those
+> used with `libcore.rlib`. In this instance, the `rlib` was compiled with
+> `-Zreg-struct-return` and this example assumes that `$cargo_flags` does not
+> pass this flag, so rustc reports a mismatch.
+>
+> Cargo can use this mechanism to determine whether it needs to rebuild the
+> standard library to ensure compatibility.
+
+The standard library will be rebuilt in its release profile and will only vary
+in the target modifier flags necessarily for it to be compatible
+([?][rationale-release-profile]).
+
+Pre-built available? | User's profile | Target modifiers changed? | Standard library re-built?
+-------------------- | -------------- | ------------------------- | --------------------------
+No                   | N/A            | N/A                       | Yes, std's `release`
+Yes                  | `dev`          | Unchanged                 | Yes, std's `release`
+Yes                  | `release`      | Unchanged                 | No
+Yes                  | `release`      | Changed                   | Yes, std's `release`
+
+*It is assumed that changing a target modifier would be part of Cargo profiles,
+hence why there is no row for "target modifiers changed, profile unchanged".*
```

Comment:
I'm having difficulty understanding this sentence in combination with the sentence below about per-target profiles not being available.

Is this stage 2 implying that it will also be adding target modifiers to the profile? Or is that a hypothetical future? If that is a new proposal for this stage, there probably should be a relatively hefty section describing the addition of a new profile field.

Are target modifiers ignored on targets that don't support that modifier?